presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/JdbcMetadata.java

```diff
@@ -107,12 +111,45 @@ public JdbcTableHandle getTableHandle(ConnectorSession session, SchemaTableName
                 handle.getCatalogName(),
                 handle.getSchemaName(),
                 handle.getTableName(),
+                handle.getColumns(),
                 newDomain,
                 handle.getLimit());
 
         return Optional.of(new ConstraintApplicationResult<>(handle, constraint.getSummary()));
     }
 
+    @Override
+    public Optional<ProjectionApplicationResult<ConnectorTableHandle>> applyProjection(
+            ConnectorSession session,
+            ConnectorTableHandle table,
+            List<ConnectorExpression> projections,
+            Map<String, ColumnHandle> assignments)
+    {
+        JdbcTableHandle handle = (JdbcTableHandle) table;
+
+        List<JdbcColumnHandle> newColumns = assignments.values().stream()
+                .map(JdbcColumnHandle.class::cast)
+                .collect(Collectors.toList());
+
+        if (handle.getColumns().isPresent() && containSameElements(newColumns, handle.getColumns().get())) {
+            return Optional.empty();
+        }
+
+        return Optional.of(new ProjectionApplicationResult<>(
+                new JdbcTableHandle(
+                        handle.getSchemaTableName(),
+                        handle.getCatalogName(),
+                        handle.getSchemaName(),
+                        handle.getTableName(),
+                        Optional.of(newColumns),
+                        handle.getConstraint(),
+                        handle.getLimit()),
+                projections,
+                assignments.entrySet().stream()
+                        .map(assignment -> new ProjectionApplicationResult.Assignment(assignment.getKey(), assignment.getValue(), ((JdbcColumnHandle) assignment.getValue()).getColumnType()))
```

Comment:
Wrapping the `Assignment` arguments would make this easier to read. Maybe also static import it.